File: 11.flood_fill/python/main.py
from typing import List, Tuple
from math import pow

def is_bound(x: int, y: int, center: List[int], radius: int) -> bool:
    x_dist = x - center[0]
    y_dist = y - center[1]
    sq_dist = int(pow(x_dist, 2)) + int(pow(y_dist, 2))
    sq_radius = int(pow(radius, 2))

    if sq_dist >= sq_radius - 20 and sq_dist <= sq_radius + 20: return True
    else: return False

def generate_circle(grid: List[List[str]], N: int):
    center: List[int] = [N//2 - N//6, N//2+N//6]
    radius: int = N//3

    for i in range(N):
        for j in range(N):
            if is_bound(i, j, center, radius): grid[i][j] = 'X'
            else: grid[i][j] = ' '

def print_grid(grid: List[List[str]], N: int):
    for g in grid:
        for c in g:
            print(c, end='')
        print('')

def valid_point(p: Tuple[int, int], grid: List[List[str]], N: int) -> bool:
    valid: bool = True

    # out of range
    if p[0]<0 or p[0]>=N or p[1]<0 or p[1]>=N: valid = False
    else:
        # already colored or is boundary
        if grid[p[0]][p[1]] == 'O' or grid[p[0]][p[1]] == 'X': valid = False

    return valid

def flood_fill(fill_point: Tuple[int, int], grid: List[List[int]], N: int):
    que: List[Tuple[int, int]] = []
    que.append(fill_point)

    while(len(que) > 0):
        p: Tuple[int, int] = que.pop(0)

        adj_points: List[Tuple[int, int]] = [(p[0], p[1]-1),(p[0], p[1]+1),(p[0]-1, p[1]),(p[0]+1, p[1])]
        for adj in adj_points:
            if valid_point(adj, grid, N):
                que.append(adj)

                # color
                grid[adj[0]][adj[1]] = 'O'

if __name__ == "__main__":
    N = 30
    grid = [[' ' for _ in range(N)] for _ in range(N)]

    generate_circle(grid, N)
    print_grid(grid, N)

    fill_point: Tuple[int, int] = (0, 0)
    flood_fill(fill_point, grid, N)
    print_grid(grid, N)

File: 11.flood_fill/cpp/src/main.cpp
#include <iostream>
#include <queue>

typedef struct Point {
    int x;
    int y;
} Point;

char** generate_circle(int N);
bool is_bound(Point p, Point center, int radius);
void print_map(char** map, int N);
void flood_fill(char** map, int N, Point fill_point);
bool valid_point(char** map, int N, Point point);

int main(int argc) {
    int N = 30;

    char** map = generate_circle(N);
    print_map(map, N);

    Point fill_point = {0, 0};

    // compute flood fill
    flood_fill(map, N, fill_point);
    print_map(map, N);

    for(int i=0;i<N;i++) {
        delete[] map[i];
    }
    delete[] map;
    return 0;
}

char** generate_circle(int N) {
    Point center = {N/2 - N/6, N/2+N/6}; 
    int radius = N/3;

    char** map = new char*[N];
    for(int i=0;i<N;i++) {
        map[i] = new char[N];
        for(int j=0;j<N;j++) {
            Point p = {i, j};
            if(is_bound(p, center, radius)) map[i][j] = 'X';
            else map[i][j] = ' ';
        }
    }

    return map;
}

bool is_bound(Point p, Point center, int radius) {
    int x_dist = p.x - center.x;
    int y_dist = p.y - center.y;
    int sq_dist = x_dist * x_dist + y_dist * y_dist;
    int sq_radius = radius * radius;

    if(sq_dist >= sq_radius-20 && sq_dist <= sq_radius+20) return true;
    else return false;
}

void print_map(char** map, int N) {
    for(int i=0;i<N;i++) {
        for(int j=0;j<N;j++) {
            printf("%c", map[i][j]);
        }
        printf("\n");
    }
}

void flood_fill(char** map, int N, Point fill_point) {
    std::queue<Point> que; 
    que.push(fill_point);

    while(!que.empty()) {
        Point p = que.front();
        que.pop();

        Point left = {p.x, p.y-1};
        Point right = {p.x, p.y+1};
        Point top = {p.x-1, p.y};
        Point bottom = {p.x+1, p.y};
        Point adj_points[] = {left, right, top, bottom};

        for(int i=0;i<4;i++) {
            if(valid_point(map, N, adj_points[i])) {
                que.push(adj_points[i]);

                // color
                map[adj_points[i].x][adj_points[i].y] = 'O';
            }
        }
    }
}

bool valid_point(char** map, int N, Point point) {
    bool valid = true;

    // out of range
    if(point.x<0||point.x>=N||point.y<0||point.y>=N) valid = false;
    else {
        // already colored or is boundary
        if(map[point.x][point.y]=='O' || map[point.x][point.y]=='X') valid = false;
    }
    return valid;
}